From 256cde4868cae676cef92c22fbd47405011a4156 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Tue, 27 Feb 2018 12:44:27 -0800
Subject: [PATCH] TRITON-143 config-agent should randomize its SAPI GetConfig
 calls Reviewed by: Todd Whiteman <todd.whiteman@joyent.com> Approved by: Todd
 Whiteman <todd.whiteman@joyent.com>

---
 Makefile     |  4 ++--
 agent.js     | 17 +++++++++++++++--
 bin/agent.sh |  9 +++++++++
 3 files changed, 26 insertions(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index 8893332..d7e3871 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -19,7 +19,7 @@ JS_FILES	:= $(shell ls *.js) $(shell find cmd lib -name '*.js')
 JSL_CONF_NODE	 = tools/jsl.node.conf
 JSL_FILES_NODE   = $(JS_FILES)
 JSSTYLE_FILES	 = $(JS_FILES)
-JSSTYLE_FLAGS    = -t 4 -o doxygen
+JSSTYLE_FLAGS    = -t 4 -o doxygen,unparenthesized-return=0
 SMF_MANIFESTS_IN = smf/manifests/config-agent.xml.in
 
 NODE_PREBUILT_VERSION=v0.10.48
diff --git a/agent.js b/agent.js
index 98ae5c1..9f94266 100644
--- a/agent.js
+++ b/agent.js
@@ -101,13 +101,26 @@ var autoMetadata = config.autoMetadata = {};
  * for config updates from SAPI and refreshing manifests/config files.
  */
 function startPeriodicRefresh() {
+	var delay;
+
+	// Return a random delay between 0.5*pollInterval and 1.5*pollInterval.
+	function getDelay() {
+		var min = config.pollInterval * 0.5;
+		var max = config.pollInterval * 1.5;
+		return Math.floor(Math.random() * (max - min + 1)) + min;
+	}
+
 	function checkOnce() {
 		agent.checkAndRefresh(function doneCheck(err) {
-			setTimeout(checkOnce, config.pollInterval);
+			delay = getDelay();
+			log.trace({delay: delay}, 'schedule checkAndRefresh');
+			setTimeout(checkOnce, delay);
 		});
 	}
 
-	setTimeout(checkOnce, config.pollInterval);
+	delay = getDelay();
+	log.trace({delay: delay}, 'schedule checkAndRefresh');
+	setTimeout(checkOnce, delay);
 }
 
 async.waterfall([
diff --git a/bin/agent.sh b/bin/agent.sh
index 9ec2ee9..194d8fc 100755
--- a/bin/agent.sh
+++ b/bin/agent.sh
@@ -116,6 +116,15 @@ if [[ -e $FIRST_RUN_FILE ]]; then
 fi
 echo "$0: start sync mode attempts (RAN_ONCE=$RAN_ONCE)"
 
+
+# Note: There is a possible dogpile-on-SAPI scenario for this first sync mode
+# config-agent run, if a large number of config-agents are restarted at the same
+# time -- e.g. all GZ config-agents for a large number of CNs.
+#
+# Adding an initial first-sync-run delay for all config-agents is undesirable
+# however, because it forever means slower Triton instance restarts.
+# Instead the expectation is that SAPI can cope with rate-limiting.
+
 COUNT=0
 SUCCESS=1
 while [[ $SUCCESS != 0 ]]; do
-- 
2.21.0

